rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,151 @@
+- Start Date: 2020-03-02
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+Currently, the applications that support screenshot reports are:
+ - Dashboard
+ - Visualize Editor
+ - Canvas
+
+Kibana UI code should be aware when the page is rendering for the purpose of
+capturing a screenshot. There should be a service to interact with low-level
+code for providing that awareness. Reporting would interact with this service
+to improve the quality of the Kibana Reporting feature for a few reasons:
+
+ - Fewer objects in the headless browser memory since interactive code doesn't run
+ - Fewer API requests made by the headless browser for features that don't apply in a non-interactive context
+
+**Screenshot mode service**
+
+The Reporting-enabled applications should use the recommended practice of
+having a customized URL for Reporting. The customized URL renders without UI
+features like navigation, auto-complete, and anything else that wouldn't make
+sense for non-interactive pages.
+
+However, applications are one piece of the UI code in a browser, and they have
+dependencies on other UI plugins. Apps can't control plugins and other things
+that Kibana loads in the browser.
+
+This RFC proposes a Screenshot Mode Service as a low-level plugin that allows
+other plugins (UI code) to make choices when the page is rendering for a screenshot.
+
+More background on how Reporting currently works, including the lifecycle of
+creating a PNG report, is here: https://github.com/elastic/kibana/issues/59396
+
+# Motivation
+
+The Reporting team wants all applications to support a customized URLs, such as
+Canvas does with its `#/export/workpad/pdf/{workpadId}` UI route. The
+customized URL is where an app can solve any rendering issue in a PDF or PNG,
+without needing extra CSS to be injected into the page.
+
+However, many low-level plugins have been added to the UI over time. These run
+on every page and an application can not turn them off. Reporting performance
+is negatively affected by this type of code. When the Reporting team analyzes
+customer logs to figure out why a job timed out, we sometimes see requests for
+the newsfeed API and telemetry API: services that aren't needed during a
+reporting job.
+
+In 7.12.0, using the customized `/export/workpad/pdf` in Canvas, the Sample
+Data Flights workpad loads 163 requests. Most of thees requests don't come from
+the app itself but from the application container code that Canvas can't turn
+off.
+
+# Detailed design
+
+The Screenshot Mode Service is an entirely new plugin that has an API method
+that returns a Boolean. The return value tells the plugin whether or not it
+should render itself to optimize for non-interactivity.
+
+The plugin is low-level as it has no dependencies of its own, so other
+low-level plugins can depend on it.
```

Comment:
Another thought here: 

We have an agreement that plugins don't manage URL and only an App should talk to the URL. 
Any state propagation should be done: URL -> app -> plugin -> app -> URL. [reasoning here](https://github.com/elastic/kibana/issues/39855)

I am curious if this is also applicable here. and the state propagation then should be something like (with an exception that `screenshot mode` plugin will read from the URL screenshot state once in setup phase:

```
screenshot mode plugin -> app -> all other plugins 
                            | -> core
```

this would also address core vs other plugins inconsistency mentioned in the comment above, without making moving `screenshot mode` to core.

Reply:
i would prefer no exception to that, the reasoning is reasonable imo :)

Reply:
I'm not sure if I follow.

Reply:
we are discussing puting a flag in the URL, and figured that would require plugin to manage the URL, but our current mentality is not to have plugins manage the URLs but have apps full responsible for it.